notebooks/2018-12-14_Fortran_Files.ipynb

```diff
@@ -0,0 +1,333 @@
+    "|------|-------|------|------|-----|    |-------|-------|-------|\n",
+    "|  2  | `int16` | `selected_int_kind(3)` |  127 | | | |\n",
+    "|  4  | `int32` | `selected_int_kind(5)` | > 2*10^9 | | | `real32` | `selected_real_kind(6, 37)` |\n",
+    "|  8  | `int64` | `selected_int_kind(10)` | > 9*10^18 | | | `real64` | `selected_real_kind(15, 307)` |\n",
+    "| 16  | ----    | `selected_int_kind(19)` | > 10^38 | | | `real128` | `selected_real_kind(33, 4931)` |\n",
+    "\n",
+    "Note that `iso_fortran_env` does not have a named type `int128`, though your compiler might have it. \n",
+    "Some compilers also have a 10-byte real kind.\n",
+    "\n",
+    "### newunit\n",
+    "\n",
+    "Whenever you interact with a file, you need a unit, an integer value that references a specific open file.\n",
+    "Some I/O streams, specifically Standard Input, Standard Output, and Standard Error have compiler dependent values for these units, which unfortunately are not standardised.\n",
+    "\n",
+    "Keeping track of these values while remaining compiler-agnostic is getting a bit confusing.\n",
+    "Fortunately, there's an option for that: `newunit`.\n",
+    "\n",
+    "Instead of using a hardcoded integer value, declare an integer variable with a meaningful name, then open the file with `newunit=` instead of `unit=` parameter:\n",
+    "\n",
+    "```fortran\n",
+    "integer :: output_handle\n",
+    "...\n",
+    "open(newunit=output_handle, file='data.dat', ...)\n",
+    "...\n",
+    "write(output_handle, *) values(:, i)\n",
+    "...\n",
+    "close(output_handle)\n",
+    "```\n",
+    "\n",
+    "A new, unused value is assigned every time you open the file, and you don't have to worry about interfering file handles any more."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Stream\n",
+    "\n",
+    "Stream output has been part of Fortran 2003 and later. \n",
+    "\n",
+    "The binary representation of the data is written directly to the file, without any metadata.\n",
+    "\n",
+    "### Fortran writing stream data\n",
+    "\n",
+    "```fortran\n",
+    "program write_stream\n",
+    "    use iso_fortran_env, only: int16\n",
+    "    implicit none\n",
+    "    integer(kind=int16) :: ii\n",
+    "    integer :: output_handle\n",
+    "    open(newunit=output_handle, file='stream_data.dat', action='write',   &\n",
+    "         status='replace', access='stream', format='unformatted')\n",
+    "    write(output_handle) [(ii, ii=1, 10)]\n",
+    "    write(output_handle) \"Hello World\"\n",
+    "    close(output_handle)\n",
+    "end program write_stream\n",
+    "``` "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "00000000  01 00 02 00 03 00 04 00  05 00 06 00 07 00 08 00  |................|\r\n",
```

Comment:
Might need to explain what we're looking at here, and why `1` is represented as `01 00`